// chokidar-watcher/dist/index.js
"use strict";
/* IMPORT */
const chokidar = require("chokidar");
const consts_1 = require("./consts");
const get_id_1 = require("./get_id");
const get_lock_1 = require("./get_lock");
/* WATCHER */
//TODO: Write a test suite
function watcher(paths, options = {}, handlers = {}) {
    /* VARIABLES */
    const locksAdd = {}, locksUnlink = {};
    /* HELPERS */
    function emit(event, args) {
        const handler = handlers[event];
        if (!handler)
            return;
        handler.apply(undefined, args);
    }
    /* HANDLERS */
    function change(filePath, stats) {
        emit('change', [filePath, stats]);
    }
    function add(filePath, stats) {
        const id = get_id_1.default(filePath, stats);
        get_lock_1.default(id, consts_1.RENAME_TIMEOUT, {
            locks: {
                read: locksUnlink,
                write: locksAdd
            },
            handlers: {
                free: () => emit('add', [filePath, stats]),
                override: (prevPath) => emit('rename', [prevPath, filePath]),
                overridden: () => filePath
            }
        });
    }
    function unlink(filePath) {
        const id = get_id_1.default(filePath);
        get_lock_1.default(id, consts_1.RENAME_TIMEOUT, {
            locks: {
                read: locksAdd,
                write: locksUnlink
            },
            handlers: {
                free: () => emit('unlink', [filePath]),
                override: (newPath) => emit('rename', [filePath, newPath]),
                overridden: () => filePath
            }
        });
    }
    /* CHOKIDAR */
    const chokidarOptions = Object.assign({}, options, { ignoreInitial: false });
    return chokidar.watch(paths, chokidarOptions).on('add', add).on('change', change).on('unlink', unlink);
}
/* EXPORT */
module.exports = watcher;
module.exports.default = watcher;
Object.defineProperty(module.exports, "__esModule", { value: true });
