// configuration/dist/utils/ajv.js
"use strict";
/* IMPORT */
Object.defineProperty(exports, "__esModule", { value: true });
/* AJV */
const AJV = {
    instance: undefined,
    getInstance() {
        if (AJV.instance)
            return AJV.instance;
        const Ajv = require('ajv'), // Lazy import for performance
        Filter = require('ajv-filter'); // Lazy import for performance
        const options = {
            sourceCode: false,
            unicode: false,
            validateSchema: false,
            serialize: false,
            cache: {
                put: () => { },
                get: () => { },
                del: () => { },
                clear: () => { }
            }
        };
        const ajv = Filter.patchInstance(new Ajv(options), true);
        return AJV.instance = ajv;
    },
    getSchema(schema) {
        const Filter = require('ajv-filter'); // Lazy import for performance
        return Filter.patchSchema(schema);
    },
    validateSchema(schema) {
        const ajv = AJV.getInstance();
        return ajv.validateSchema(schema);
    },
    getValidator(schema) {
        const ajv = AJV.getInstance();
        return ajv.compile(schema);
    }
};
/* EXPORT */
exports.default = AJV;
